File: flutter_app/lib/domain/calculations/cycle_adaptation_policy.dart
import '../enums.dart';
import '../models/program_models.dart';

enum AdaptationReadinessTier { low, neutral, high }

enum AdaptationConfidence { low, medium, high }

class CycleAdaptationPolicy {
  const CycleAdaptationPolicy();

  static const Map<CyclePhase, ProgramPhaseAdjustmentSettings>
      _baselinePhaseSettings = <CyclePhase, ProgramPhaseAdjustmentSettings>{
    CyclePhase.menstrual: ProgramPhaseAdjustmentSettings(
      loadMultiplier: 0.9,
      setsMultiplier: 0.9,
      repsMultiplier: 0.9,
    ),
    CyclePhase.follicular: ProgramPhaseAdjustmentSettings(
      loadMultiplier: 1.0,
      setsMultiplier: 1.0,
      repsMultiplier: 1.0,
    ),
    CyclePhase.ovulation: ProgramPhaseAdjustmentSettings(
      loadMultiplier: 1.12,
      setsMultiplier: 1.05,
      repsMultiplier: 0.95,
    ),
    CyclePhase.luteal: ProgramPhaseAdjustmentSettings(
      loadMultiplier: 0.97,
      setsMultiplier: 0.95,
      repsMultiplier: 0.92,
    ),
  };

  static const Map<AdaptationReadinessTier, double> _readinessScales =
      <AdaptationReadinessTier, double>{
    AdaptationReadinessTier.low: 0.6,
    AdaptationReadinessTier.neutral: 1.0,
    AdaptationReadinessTier.high: 1.2,
  };

  static const Map<AdaptationConfidence, double> _confidenceScales =
      <AdaptationConfidence, double>{
    AdaptationConfidence.low: 0.55,
    AdaptationConfidence.medium: 0.8,
    AdaptationConfidence.high: 1.0,
  };

  AdaptationReadinessTier resolveReadinessTier({double? readinessScore}) {
    if (readinessScore == null) {
      return AdaptationReadinessTier.neutral;
    }
    if (readinessScore <= 3) {
      return AdaptationReadinessTier.low;
    }
    if (readinessScore >= 8) {
      return AdaptationReadinessTier.high;
    }
    return AdaptationReadinessTier.neutral;
  }

  double resolveConfidenceScale({
    required AdaptationConfidence confidence,
    ProgramCycleAdjustmentProfile? profile,
  }) {
    final double base = _confidenceScales[confidence] ?? 1.0;
    if (confidence != AdaptationConfidence.low) {
      return base;
    }

    final double profileScale = profile?.lowConfidenceScale ?? 0.7;
    return (base * profileScale).clamp(0.1, 1.0);
  }

  CyclePhase resolvePhase({
    required CyclePhase? currentPhase,
    required CyclePhase? lastKnownPhase,
    ProgramCycleAdjustmentProfile? profile,
  }) {
    if (currentPhase != null) {
      return currentPhase;
    }
    if (lastKnownPhase != null) {
      return lastKnownPhase;
    }
    final String fallback = profile?.fallbackPhase ?? 'follicular';
    return _phaseFromRaw(fallback) ?? CyclePhase.follicular;
  }

  AdaptationConfidence resolveConfidence({
    required CyclePhase? currentPhase,
    required CyclePhase? lastKnownPhase,
    required int periodLogCount,
    required DateTime? lastPeriodStart,
    DateTime? referenceDate,
  }) {
    if (periodLogCount == 0) {
      return AdaptationConfidence.low;
    }

    final DateTime now = referenceDate ?? DateTime.now();
    if (lastPeriodStart != null &&
        now.difference(lastPeriodStart).inDays > 45) {
      return AdaptationConfidence.low;
    }

    if (currentPhase != null && periodLogCount >= 3) {
      return AdaptationConfidence.high;
    }

    if (currentPhase != null) {
      return AdaptationConfidence.medium;
    }

    if (lastKnownPhase != null) {
      return AdaptationConfidence.low;
    }

    return AdaptationConfidence.low;
  }

  ProgramExercise applyPhaseAdjustment({
    required ProgramExercise exercise,
    required CyclePhase phase,
    required AdaptationReadinessTier readinessTier,
    required AdaptationConfidence confidence,
    required ProgramCycleAdjustmentProfile? profile,
  }) {
    final ProgramPhaseAdjustmentSettings targetSettings =
        _resolvePhaseSettings(phase: phase, profile: profile);
    final double readinessScale = _readinessScales[readinessTier] ?? 1.0;
    final double confidenceScale = resolveConfidenceScale(
      confidence: confidence,
      profile: profile,
    );

    final double loadMultiplier = _blendMultiplier(
      targetMultiplier: targetSettings.loadMultiplier,
      readinessScale: readinessScale,
      confidenceScale: confidenceScale,
    );
    final double setsMultiplier = _blendMultiplier(
      targetMultiplier: targetSettings.setsMultiplier,
      readinessScale: readinessScale,
      confidenceScale: confidenceScale,
    );
    final double repsMultiplier = _blendMultiplier(
      targetMultiplier: targetSettings.repsMultiplier,
      readinessScale: readinessScale,
      confidenceScale: confidenceScale,
    );

    return ProgramExercise(
      exercise: exercise.exercise,
      variant: exercise.variant,
      sets: _scaleExerciseValue(exercise.sets, setsMultiplier),
      reps: _scaleExerciseValue(exercise.reps, repsMultiplier),
      percent1Rm: exercise.percent1Rm == null
          ? null
          : _clampPercent1Rm(exercise.percent1Rm! * loadMultiplier),
      restMinutes: exercise.restMinutes,
      notes: exercise.notes,
    );
  }

  ProgramPhaseAdjustmentSettings _resolvePhaseSettings({
    required CyclePhase phase,
    required ProgramCycleAdjustmentProfile? profile,
  }) {
    final ProgramPhaseAdjustmentSettings? profileSetting =
        profile?.phaseSettings[phase.name];
    return profileSetting ?? _baselinePhaseSettings[phase]!;
  }

  double _blendMultiplier({
    required double targetMultiplier,
    required double readinessScale,
    required double confidenceScale,
  }) {
    final double direction = targetMultiplier - 1.0;
    final double weighted =
        1.0 + (direction * readinessScale * confidenceScale);
    return weighted.clamp(0.75, 1.25);
  }

  ExerciseValue _scaleExerciseValue(ExerciseValue value, double multiplier) {
    switch (value.type) {
      case ExerciseValueType.fixed:
        final int base = value.fixedValue ?? 1;
        final int adjusted = (base * multiplier).round().clamp(1, 20);
        return ExerciseValue.fixed(adjusted);
      case ExerciseValueType.range:
        final int baseMin = value.minValue ?? 1;
        final int baseMax = value.maxValue ?? baseMin;
        final int adjustedMin = (baseMin * multiplier).round().clamp(1, 20);
        final int adjustedMax = (baseMax * multiplier).round().clamp(1, 20);
        final int minValue =
            adjustedMin <= adjustedMax ? adjustedMin : adjustedMax;
        final int maxValue =
            adjustedMax >= adjustedMin ? adjustedMax : adjustedMin;
        return ExerciseValue.range(minValue, maxValue);
      case ExerciseValueType.amrap:
      case ExerciseValueType.text:
        return value;
    }
  }

  double _clampPercent1Rm(double value) {
    return value.clamp(0.4, 1.1);
  }

  CyclePhase? _phaseFromRaw(String raw) {
    for (final CyclePhase phase in CyclePhase.values) {
      if (phase.name == raw) {
        return phase;
      }
    }
    return null;
  }
}

File: android/app/src/main/java/com/sundeefundee/domain/CycleAdaptationPolicy.kt
package com.sundeefundee.domain

/**
 * Adaptation readiness tiers based on user-reported energy/fatigue scores.
 */
enum class AdaptationReadinessTier {
    LOW,
    NEUTRAL,
    HIGH
}

/**
 * Confidence levels for cycle phase predictions.
 */
enum class AdaptationConfidence {
    LOW,
    MEDIUM,
    HIGH
}

/**
 * Phase-specific baseline settings for load, sets, and reps multipliers.
 */
data class PhaseSettings(
    val loadMultiplier: Double,
    val setsMultiplier: Double,
    val repsMultiplier: Double
)

/**
 * Policy for adapting workout programs based on menstrual cycle phase.
 * Pure value semantics - never mutates inputs, returns new instances.
 */
object CycleAdaptationPolicy {

    private val baselinePhaseSettings: Map<CyclePhase, PhaseSettings> = mapOf(
        CyclePhase.MENSTRUAL to PhaseSettings(0.90, 0.90, 0.90),
        CyclePhase.FOLLICULAR to PhaseSettings(1.00, 1.00, 1.00),
        CyclePhase.OVULATION to PhaseSettings(1.12, 1.05, 0.95),
        CyclePhase.LUTEAL to PhaseSettings(0.97, 0.95, 0.92)
    )

    private val readinessScales: Map<AdaptationReadinessTier, Double> = mapOf(
        AdaptationReadinessTier.LOW to 0.6,
        AdaptationReadinessTier.NEUTRAL to 1.0,
        AdaptationReadinessTier.HIGH to 1.2
    )

    private val confidenceScales: Map<AdaptationConfidence, Double> = mapOf(
        AdaptationConfidence.LOW to 0.55,
        AdaptationConfidence.MEDIUM to 0.8,
        AdaptationConfidence.HIGH to 1.0
    )

    /**
     * Determines readiness tier from an optional readiness score (1-10 scale).
     */
    fun resolveReadinessTier(readinessScore: Int?): AdaptationReadinessTier {
        if (readinessScore == null) return AdaptationReadinessTier.NEUTRAL
        return when {
            readinessScore <= 3 -> AdaptationReadinessTier.LOW
            readinessScore >= 8 -> AdaptationReadinessTier.HIGH
            else -> AdaptationReadinessTier.NEUTRAL
        }
    }

    /**
     * Resolves confidence scale based on confidence level.
     */
    fun resolveConfidenceScale(
        confidence: AdaptationConfidence,
        lowConfidenceScale: Double = 0.7
    ): Double {
        val base = confidenceScales[confidence] ?: 1.0
        return if (confidence == AdaptationConfidence.LOW) {
            (base * lowConfidenceScale).coerceIn(0.1, 1.0)
        } else {
            base
        }
    }

    /**
     * Resolves the current cycle phase with fallback logic.
     */
    fun resolvePhase(
        currentPhase: CyclePhase?,
        lastKnownPhase: CyclePhase?,
        fallbackPhase: CyclePhase = CyclePhase.FOLLICULAR
    ): CyclePhase {
        return currentPhase ?: lastKnownPhase ?: fallbackPhase
    }

    /**
     * Determines confidence level based on available data.
     */
    fun resolveConfidence(
        currentPhase: CyclePhase?,
        lastPeriodStart: Long?,
        periodLogCount: Int,
        referenceDate: Long = System.currentTimeMillis()
    ): AdaptationConfidence {
        if (periodLogCount <= 0) return AdaptationConfidence.LOW

        if (lastPeriodStart != null) {
            val daysSinceLastPeriod = (referenceDate - lastPeriodStart) / (24 * 60 * 60 * 1000)
            if (daysSinceLastPeriod > 45) {
                return AdaptationConfidence.LOW
            }
        }

        return when {
            currentPhase != null && periodLogCount >= 3 -> AdaptationConfidence.HIGH
            currentPhase != null -> AdaptationConfidence.MEDIUM
            else -> AdaptationConfidence.LOW
        }
    }

    /**
     * Applies phase adjustment to a target multiplier.
     */
    fun applyPhaseAdjustment(
        targetMultiplier: Double,
        readinessTier: AdaptationReadinessTier,
        confidence: AdaptationConfidence
    ): Double {
        val rs = readinessScales[readinessTier] ?: 1.0
        val cs = resolveConfidenceScale(confidence)
        val blended = 1.0 + (targetMultiplier - 1.0) * rs * cs
        return blended.coerceIn(0.75, 1.25)
    }

    /**
     * Gets phase settings for a given cycle phase.
     */
    fun getPhaseSettings(phase: CyclePhase): PhaseSettings {
        return baselinePhaseSettings[phase] ?: PhaseSettings(1.0, 1.0, 1.0)
    }

    /**
     * Gets load multiplier for a given cycle phase.
     */
    fun getLoadMultiplier(phase: CyclePhase): Double {
        return getPhaseSettings(phase).loadMultiplier
    }

    /**
     * Gets sets multiplier for a given cycle phase.
     */
    fun getSetsMultiplier(phase: CyclePhase): Double {
        return getPhaseSettings(phase).setsMultiplier
    }

    /**
     * Gets reps multiplier for a given cycle phase.
     */
    fun getRepsMultiplier(phase: CyclePhase): Double {
        return getPhaseSettings(phase).repsMultiplier
    }
}

/**
 * Extension function to clamp a comparable value within a range.
 */
fun <T : Comparable<T>> T.clamped(to range: ClosedRange<T>): T {
    return maxOf(minOf(this, range.endInclusive), range.start)
}
